### ISSUE TITLE
More aggresive optimizations using ".output"

### ISSUE THREAD

**Issue by Escapingbug** [souffle-lang/souffle#1498]
On this snippet:

```datalog
.decl A(x:number)

A(0).
A(x + 1) :- A(x).

.decl B(x:number)

B(i) :- A(i), i <= 10.

.output B(io=stdout)
```

What I expected is that this program should terminate and output B(i) = A(0) where i <= 10, but actually this one never terminates.

Since the output denotes what I wanted, and by that we can infer that only first 10 facts are used, maybe we are able to make this happen?

Also, things like this might reduce memory usage, since this will generate less facts than before.

**Comment by azreika** [comment]
A few running thinking notes. The underlying idea here is to restrict intermediate relations (like `A` here) to only produce tuples that contribute to the output.

In this particular example, we would want `A` to be restricted by the `i <= 10` constraint that appears in the only place it's used. Inlining can't be used here, since `A` is self-recursive, but we can rewrite the program in a similar way by pushing up the constraint:

```
.decl A(x:number)

A(0).
A(x + 1) :- A(x), x <= 10.

.decl B(x:number)

B(i) :- A(i), i <= 10.

.output B()
```
or, equivalently, since `B` is now (almost) a copy of `A`:
```
.decl B(x:number)
B(0).
B(x) :- A(x), x <= 10.
.output B()
```
Unfortunately, this kind of transformation is not always possible. For example, suppose the constraint was `x % 2 == 0`, then the correct output SHOULD be infinite (every even number). But, if we do the same sort of transformation, we will only be producing the number `0`.
```
.decl A(x:number)
A(0).
A(x + 1) :- A(x), x % 2 == 0. // only produces {1} now, since 1 % 2 != 0

.decl B(x:number)
B(i) :- A(i), i % 2 == 0.

.output B()
```

As a more restrictive example, consider:
```
.decl A(x:number)

A(0).
A(x + 2) :- A(x).

.decl B(x:number)

B(i) :- A(i), i == 100.

.output B()
```
Now we need to produce 100 other tuples before we find out `A` contains 100, so when would we know when to stop looking?

**Comment by Escapingbug** [comment]
I think I've get the most of the ideas here. And I know it must contains some corner cases that are hard to resolve, thanks for replying me with that, haha.

But there's also one thing I don't understand so much. Why must souffle use a "bottom-up" approach of generating facts? Since as a quick thinking, a top-down approach should not have this kind of problem? (I'm really new to this area, so this might be a really silly question.)

**Comment by azreika** [comment]
Not a silly question at all!  Top-down (like Prolog) and bottom-up (like Souffle's Datalog) each have their pros and cons, and behave very differently depending on your problem and coding style. You're right in that bottom-up is less "goal-oriented", and so it's more likely to generate tuples that aren't relevant to the query (like in your original program). However, top-down suffers from (sometimes exponential) blow-up in intermediate proof trees, and is more likely to generate repeated facts.

For Souffle's use-cases, we typically expect to generate a huge amount of output tuples. For this purpose, top-down evaluation tends to behave much worse.

We currently have a set of techniques to make the bottom-up approach more goal-oriented, so that we can get the best of both worlds. For example, the magic-set transformation, which is currently in the process of being fixed up (issue #1491). The magic-set transformation pushes up constraints/requirements from the goal into the intermediate rules of the program, so a tighter set of relevant tuples are ever added. Techniques like this tend to work fairly well in optimisations. [Some papers](https://dl.acm.org/doi/10.1145/73721.73736) also show that for certain classes of Datalog programs evaluated using a particular top-down approach, there always exists a rewriting of the program that performs the same or better using semi-naive bottom-up evaluation.

**Comment by Escapingbug** [comment]
Thanks for such detailed reply. I truly learned a lot from it! I think maybe magic-set transformation could help. So I'll close this issue.
